[RFC] test for PR #4646

justinmk opened the pull request:
cc @oni-link 

~~The test fails correctly:~~ **edit:** This isn't failing in CI. Hmm...

```
Expected:
|*bytes: 125769                                     |
|     1 HelloWorld:qa                              |
|  5989 HelloWorld:qa                              |
|~                                                 |
|[No Name] [+]                                     |
|                                                  |
|-- TERMINAL --                                    |
Actual:
|*bytes: 65519                                      |
|     1 HelloWorld:qa                              |
|  3120 HelloWorld:qa                              |
|~                                                 |
|[No Name] [+]                                     |
|                                                  |
|-- TERMINAL --                                    |
```
- Note: the test does not _always_ fail, but it _usually_ fails. 
  - Retrying the test 3 times brings failure close to 100%.
- There shouldn't be a false-positive, because the test condition happens in the `on_exit` handler.

justinmk committed:
test/tui/job_spec.lua

justinmk commented on test/functional/terminal/job_spec.lua:
```diff
@@ -0,0 +1,71 @@
+      thelpers.feed_data(":call jobstart(['bash', '-c', 'cat "..input_file.."'], g:callbacks)\n")
+    end
+
+    helpers.sleep(3000)  -- Give the job time to finish.
```
I'll change this to something a bit smarter.

justinmk commented on test/functional/terminal/job_spec.lua:
```diff
@@ -0,0 +1,71 @@
+    thelpers.feed_data(':source Xtest_job_4569.vim\n')
+
+    if helpers.os_name == 'windows' then
+      thelpers.feed_data(":call jobstart(['powershell', '-c', 'cat \""..input_file.."\"], g:callbacks)\n")
```
powershell tends to have reasonable aliases to common unix shell commands (cat, grep, echo), so that should help with porting Windows tests.

justinmk committed:
test/tui/job_spec.lua: Remove byte-count assert.

This is unreliable by ~3 bytes--possibly due to shell output variation?

justinmk committed:
test/tui/job_spec.lua: Run 3 times.

Before the bugfix, the test fails roughly 40% of the time.
Running it 3 times brings it close to 100%.

justinmk commented on test/functional/terminal/job_spec.lua:
```diff
@@ -0,0 +1,71 @@
+    '", "-u", "NONE", "-i", "NONE", "--cmd", "set noswapfile"]')
+    screen:set_default_attr_ids({})
+    screen:set_default_attr_ignore(true)
+    screen.timeout = 15000
```
This is a max of 15 seconds, but with the bugfix from #4646 the test succeeds in `.22` seconds (even with 3 retries).

justinmk commented on test/functional/terminal/job_spec.lua:
```diff
@@ -0,0 +1,71 @@
+  it('does not lose data (#4569)', function()
+    test_job_output_collection()
+    test_job_output_collection()
+    test_job_output_collection()
```
Retry 3 times to avoid false-negatives.

justinmk commented on test/functional/terminal/job_spec.lua:
```diff
@@ -31,9 +30,6 @@ describe('job handler', function()
           put =join(a:data,"\n")
         elseif 'exit' ==# a:event
-          " Length of received data.
-          call setline('.', 'bytes: '.len(join(g:job_stdout[0:-2],"\n")))
```
@oni-link This was unreliable, even after merging #4646. It varied by about 3 bytes. Is it fair to say this could be due to shell output inconsistencies?

jamessan replied:
Why would there be inconsistencies in the output of a non-interactive shell that just cats a file?

justinmk replied:
@jamessan I don't know :( More likely it's some sort of off-by-1 problem in nvim.

justinmk replied:
@jamessan There might not be a bug here after all. Instead this might be the result of the behavior described here: https://github.com/neovim/neovim/issues/4897#issuecomment-224754553

Because the job handlers here are blindly joining `a:data` without regards to whether `a:data` continues a previous partial result.

jamessan replied:
> Because the job handlers here are blindly joining a:data without regards to whether a:data continues a previous partial result.

Then the help should be clarified.  I've read ZyX's response there and the help a few times now and I can't fully understand ZyX's description nor see anything in the help that agrees with what I think he's trying to state.

justinmk committed:
test/tui/job_spec.lua: Add another test.

justinmk commented:
QB ubuntu 64-bit build (which matches the VM where the test also fails for me) fails correctly with the latest additional test (112b8bb):

```
19:34:24,350 INFO  - # test/functional/terminal/job_spec.lua @ 130
19:34:24,350 INFO  - # Failure message: ./test/functional/ui/screen.lua:289: Row 1 didn't match.
19:34:24,350 INFO  - # Expected:
19:34:24,350 INFO  - # |*unicode/UnicodeData.txt:10FFFD;<Plane 16 Private U|
19:34:24,350 INFO  - # |se, Last>;Co;0;L;;;;;N;;;;; |
19:34:24,350 INFO  - # | |
19:34:24,350 INFO  - # |~ |
19:34:24,350 INFO  - # |[No Name] [+] |
19:34:24,350 INFO  - # | |
19:34:24,350 INFO  - # |-- TERMINAL -- |
19:34:24,350 INFO  - # Actual:
19:34:24,350 INFO  - # |*unicode/UnicodeData.txt:2F870;CJK COMPATIBILITY ID|
19:34:24,350 INFO  - # |EOGRAPH-2F870;Lo;0;L;5BF3;;;;N;;;;; |
19:34:24,350 INFO  - # |unicode/UnicodeData.txt:2F871;CJK COMPATIBILITY ID|
19:34:24,350 INFO  - # |EOGRAPH-2F87 |
19:34:24,350 INFO  - # |[No Name] [+] |
19:34:24,350 INFO  - # | |
19:34:24,350 INFO  - # |-- TERMINAL -- |
19:34:24,350 INFO  - # stack traceback:
19:34:24,350 INFO  - #     ./test/functional/ui/screen.lua:289: in function 'wait'
19:34:24,350 INFO  - #     ./test/functional/ui/screen.lua:221: in function 'expect'
19:34:24,350 INFO  - #     test/functional/terminal/job_spec.lua:119: in function 'test_job_output_collection'
19:34:24,350 INFO  - #     test/functional/terminal/job_spec.lua:131: in function <test/functional/terminal/job_spec.lua:130>
19:34:24,350 INFO  - #
```

Travis still passes all builds though. At least I'd say the test is worth including, though it unfortunately is very system-dependent.

justinmk committed:
test/tui/job_spec.lua: WIP

justinmk committed:
test/tui/job_spec.lua: WIP

justinmk commented:
Latest version, both tests fail correctly in QB ubuntu 64. Here's job_spec.lua:

```
01:38:21,257 INFO  - not ok 1549 - job handler does not lose data (#4569)
01:38:21,257 INFO  - # test/functional/terminal/job_spec.lua @ 62
01:38:21,257 INFO  - # Failure message: ./test/functional/ui/screen.lua:289: Row 3 didn't match.
01:38:21,257 INFO  - # Expected:
01:38:21,257 INFO  - # | |
01:38:21,257 INFO  - # | 1 HelloWorld:qa |
01:38:21,257 INFO  - # |* 5989 HelloWorld:qa |
01:38:21,257 INFO  - # |~ |
01:38:21,257 INFO  - # |[No Name] [+] |
01:38:21,257 INFO  - # |"Xtest_job_4569.vim" 20L, 687C |
01:38:21,257 INFO  - # |-- TERMINAL -- |
01:38:21,257 INFO  - # Actual:
01:38:21,257 INFO  - # | |
01:38:21,257 INFO  - # | 1 HelloWorld:qa |
01:38:21,257 INFO  - # |* 3120 HelloWorld:qa |
01:38:21,257 INFO  - # |~ |
01:38:21,257 INFO  - # |[No Name] [+] |
01:38:21,257 INFO  - # |"Xtest_job_4569.vim" 20L, 687C |
01:38:21,257 INFO  - # |-- TERMINAL -- |
```

justinmk committed:
test/tui/job_spec.lua: do more work

The VimL work included in the same user command affects the outcome of
the test.

justinmk commented:
https://github.com/neovim/neovim/issues/3772#issuecomment-162334005 might be a more reliable test.
